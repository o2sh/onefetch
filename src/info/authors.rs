use super::git::sig::Sig;
use crate::{
    cli::NumberSeparator,
    info::utils::{format_number, info_field::InfoField},
};
use serde::Serialize;
use std::{collections::HashMap, fmt::Write};

#[derive(Serialize, Clone, Debug, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct Author {
    pub name: String,
    email: Option<String>,
    nbr_of_commits: usize,
    contribution: usize,
    #[serde(skip_serializing)]
    number_separator: NumberSeparator,
}

impl Author {
    pub fn new(
        name: String,
        email: Option<String>,
        nbr_of_commits: usize,
        total_nbr_of_commits: usize,
        number_separator: NumberSeparator,
    ) -> Self {
        let contribution =
            (nbr_of_commits as f32 * 100. / total_nbr_of_commits as f32).round() as usize;
        Self {
            name,
            email,
            nbr_of_commits,
            contribution,
            number_separator,
        }
    }
}

impl std::fmt::Display for Author {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        if let Some(email) = &self.email {
            write!(
                f,
                "{}% {} <{}> {}",
                self.contribution,
                self.name,
                email,
                format_number(&self.nbr_of_commits, self.number_separator)
            )
        } else {
            write!(
                f,
                "{}% {} {}",
                self.contribution,
                self.name,
                format_number(&self.nbr_of_commits, self.number_separator)
            )
        }
    }
}

#[derive(Serialize)]
pub struct AuthorsInfo {
    pub authors: Vec<Author>,
}

impl AuthorsInfo {
    pub fn new(
        number_of_commits_by_signature: &HashMap<Sig, usize>,
        total_number_of_commits: usize,
        number_of_authors_to_display: usize,
        show_email: bool,
        number_separator: NumberSeparator,
    ) -> Self {
        let authors = compute_authors(
            number_of_commits_by_signature,
            total_number_of_commits,
            number_of_authors_to_display,
            show_email,
            number_separator,
        );
        Self { authors }
    }

    fn top_contribution(&self) -> usize {
        if let Some(top_contributor) = self.authors.first() {
            return top_contributor.contribution;
        }
        0
    }
}

fn compute_authors(
    number_of_commits_by_signature: &HashMap<Sig, usize>,
    total_number_of_commits: usize,
    number_of_authors_to_display: usize,
    show_email: bool,
    number_separator: NumberSeparator,
) -> Vec<Author> {
    let mut signature_with_number_of_commits_sorted: Vec<(&Sig, &usize)> =
        Vec::from_iter(number_of_commits_by_signature);

    signature_with_number_of_commits_sorted.sort_by(|(sa, a_count), (sb, b_count)| {
        b_count.cmp(a_count).then_with(|| sa.name.cmp(&sb.name))
    });

    let authors: Vec<Author> = signature_with_number_of_commits_sorted
        .into_iter()
        .map(|(author, author_nbr_of_commits)| {
            Author::new(
                author.name.to_string(),
                if show_email {
                    Some(author.email.to_string())
                } else {
                    None
                },
                *author_nbr_of_commits,
                total_number_of_commits,
                number_separator,
            )
        })
        .take(number_of_authors_to_display)
        .collect();
    authors
}

fn digit_difference(num1: usize, num2: usize) -> usize {
    let count_digits = |num: usize| (num.checked_ilog10().unwrap_or(0) + 1) as usize;
    count_digits(num1).abs_diff(count_digits(num2))
}

impl std::fmt::Display for AuthorsInfo {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        let mut authors_info = String::new();

        let pad = self.title().len() + 2;
        for (i, author) in self.authors.iter().enumerate() {
            if i == 0 {
                write!(authors_info, "{author}")?;
            } else {
                write!(
                    authors_info,
                    "\n{:<width$}{}",
                    "",
                    author,
                    width = pad + digit_difference(self.top_contribution(), author.contribution)
                )?;
            }
        }

        write!(f, "{authors_info}")
    }
}

#[typetag::serialize]
impl InfoField for AuthorsInfo {
    fn value(&self) -> String {
        self.to_string()
    }

    fn title(&self) -> String {
        let mut title: String = "Author".into();
        if self.authors.len() > 1 {
            title.push('s')
        }
        title
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::ui::text_colors::TextColors;
    use insta::assert_snapshot;
    use owo_colors::DynColors;
    use rstest::rstest;

    #[test]
    fn test_display_author() {
        let author = Author::new(
            "John Doe".into(),
            Some("john.doe@email.com".into()),
            1500,
            2000,
            NumberSeparator::Plain,
        );

        assert_eq!(author.to_string(), "75% John Doe <john.doe@email.com> 1500");
    }

    #[test]
    fn test_display_author_with_no_email() {
        let author = Author::new("John Doe".into(), None, 1500, 2000, NumberSeparator::Plain);

        assert_eq!(author.to_string(), "75% John Doe 1500");
    }

    #[test]
    fn test_authors_info_title_with_one_author() {
        let author = Author::new(
            "John Doe".into(),
            Some("john.doe@email.com".into()),
            1500,
            2000,
            NumberSeparator::Plain,
        );

        let authors_info = AuthorsInfo {
            authors: vec![author],
        };

        assert_eq!(authors_info.title(), "Author");
    }

    #[test]
    fn test_authors_info_title_with_two_authors() {
        let author = Author::new(
            "John Doe".into(),
            Some("john.doe@email.com".into()),
            1500,
            2000,
            NumberSeparator::Plain,
        );

        let author_2 = Author::new(
            "Roberto Berto".into(),
            None,
            240,
            300,
            NumberSeparator::Plain,
        );

        let authors_info = AuthorsInfo {
            authors: vec![author, author_2],
        };

        assert_eq!(authors_info.title(), "Authors");
    }

    #[test]
    fn test_author_info_with_one_author() {
        let author = Author::new(
            "John Doe".into(),
            Some("john.doe@email.com".into()),
            1500,
            2000,
            NumberSeparator::Plain,
        );

        let authors_info = AuthorsInfo {
            authors: vec![author],
        };
        let colors = TextColors::new(&[], DynColors::Rgb(0xFF, 0xFF, 0xFF));
        let mut buffer = String::new();
        authors_info
            .write_styled(&mut buffer, false, &colors)
            .unwrap();

        assert_snapshot!(buffer);
    }

    #[test]
    fn test_author_info_with_two_authors() {
        let author = Author::new(
            "John Doe".into(),
            Some("john.doe@email.com".into()),
            1500,
            2000,
            NumberSeparator::Plain,
        );

        let author_2 = Author::new(
            "Roberto Berto".into(),
            None,
            240,
            300,
            NumberSeparator::Plain,
        );

        let authors_info = AuthorsInfo {
            authors: vec![author, author_2],
        };

        let colors = TextColors::new(&[], DynColors::Rgb(0xFF, 0xFF, 0xFF));
        let mut buffer = String::new();
        authors_info
            .write_styled(&mut buffer, false, &colors)
            .unwrap();

        assert_snapshot!(buffer);
    }
    #[test]
    fn test_author_info_alignment_with_three_authors() {
        let author = Author::new(
            "John Doe".into(),
            Some("john.doe@email.com".into()),
            1500,
            2000,
            NumberSeparator::Plain,
        );

        let author_2 = Author::new(
            "Roberto Berto".into(),
            None,
            240,
            300,
            NumberSeparator::Plain,
        );

        let author_3 = Author::new("Jane Doe".into(), None, 1, 100, NumberSeparator::Plain);

        let authors_info = AuthorsInfo {
            authors: vec![author, author_2, author_3],
        };

        let colors = TextColors::new(&[], DynColors::Rgb(0xFF, 0xFF, 0xFF));
        let mut buffer = String::new();
        authors_info
            .write_styled(&mut buffer, false, &colors)
            .unwrap();

        assert_snapshot!(buffer);
    }

    #[rstest]
    #[case(456, 123, 0)]
    #[case(456789, 123, 3)]
    #[case(1, 12, 1)]
    fn test_digit_difference(#[case] num1: usize, #[case] num2: usize, #[case] expected: usize) {
        let result = digit_difference(num1, num2);
        assert_eq!(result, expected);
    }

    #[test]
    fn test_compute_authors() {
        let mut number_of_commits_by_signature: HashMap<Sig, usize> = HashMap::new();
        number_of_commits_by_signature.insert(
            Sig {
                name: "John Doe".into(),
                email: "johndoe@example.com".into(),
            },
            30,
        );
        number_of_commits_by_signature.insert(
            Sig {
                name: "Jane Doe".into(),
                email: "janedoe@example.com".into(),
            },
            20,
        );
        number_of_commits_by_signature.insert(
            Sig {
                name: "Ellen Smith".into(),
                email: "ellensmith@example.com".into(),
            },
            50,
        );
        let total_number_of_commits = 100;
        let number_of_authors_to_display = 2;
        let show_email = false;
        let number_separator = NumberSeparator::Comma;

        let actual = compute_authors(
            &number_of_commits_by_signature,
            total_number_of_commits,
            number_of_authors_to_display,
            show_email,
            number_separator,
        );

        let expected = vec![
            Author::new(String::from("Ellen Smith"), None, 50, 100, number_separator),
            Author::new(String::from("John Doe"), None, 30, 100, number_separator),
        ];
        assert_eq!(actual, expected);
    }
}
