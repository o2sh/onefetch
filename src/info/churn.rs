use super::utils::{git::CommitMetrics, info_field::InfoField};
use crate::{cli::NumberSeparator, info::format_number};
use owo_colors::{DynColors, OwoColorize};
use serde::Serialize;
use std::fmt::Write;

#[derive(Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct FileChurn {
    pub file_path: String,
    pub nbr_of_commits: usize,
    #[serde(skip_serializing)]
    number_separator: NumberSeparator,
}

impl FileChurn {
    pub fn new(
        file_path: String,
        nbr_of_commits: usize,
        number_separator: NumberSeparator,
    ) -> Self {
        Self {
            file_path,
            nbr_of_commits,
            number_separator,
        }
    }
}

impl std::fmt::Display for FileChurn {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(
            f,
            "{} {}",
            shorten_file_path(&self.file_path, 2),
            format_number(&self.nbr_of_commits, self.number_separator)
        )
    }
}

#[derive(Serialize)]
pub struct ChurnInfo {
    pub file_churns: Vec<FileChurn>,
    pub churn_pool_size: usize,
    #[serde(skip_serializing)]
    pub info_color: DynColors,
}
impl ChurnInfo {
    pub fn new(info_color: DynColors, commit_metrics: &CommitMetrics) -> Self {
        let file_churns = commit_metrics.file_churns_to_display.clone();
        Self {
            file_churns,
            churn_pool_size: commit_metrics.churn_pool_size,
            info_color,
        }
    }
}
impl std::fmt::Display for ChurnInfo {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        let mut churn_info = String::new();

        let pad = self.title().len() + 2;

        for (i, churn) in self.file_churns.iter().enumerate() {
            let churn_str = churn.color(self.info_color);

            if i == 0 {
                write!(churn_info, "{churn_str}")?;
            } else {
                write!(churn_info, "\n{:<width$}{}", "", churn_str, width = pad)?;
            }
        }

        write!(f, "{churn_info}")
    }
}

#[typetag::serialize]
impl InfoField for ChurnInfo {
    fn value(&self) -> String {
        self.to_string()
    }

    fn title(&self) -> String {
        format!("Churn ({})", self.churn_pool_size)
    }

    fn should_color(&self) -> bool {
        false
    }
}

fn shorten_file_path(file_path: &str, depth: usize) -> String {
    let components: Vec<&str> = file_path.split('/').collect();

    if depth == 0 || components.len() <= depth {
        return file_path.to_string();
    }

    let truncated_components: Vec<&str> = components
        .iter()
        .skip(components.len() - depth)
        .copied()
        .collect();

    format!("\u{2026}/{}", truncated_components.join("/"))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_display_file_churn() {
        let file_churn = FileChurn::new("path/to/file.txt".into(), 50, NumberSeparator::Plain);

        assert_eq!(file_churn.to_string(), "\u{2026}/to/file.txt 50");
    }

    #[test]
    fn test_churn_info_value_with_two_file_churns() {
        let file_churn_1 = FileChurn::new("path/to/file.txt".into(), 50, NumberSeparator::Plain);
        let file_churn_2 = FileChurn::new("file_2.txt".into(), 30, NumberSeparator::Plain);

        let churn_info = ChurnInfo {
            file_churns: vec![file_churn_1, file_churn_2],
            churn_pool_size: 5,
            info_color: DynColors::Rgb(255, 0, 0),
        };

        assert!(churn_info.value().contains(
            &"\u{2026}/to/file.txt 50"
                .color(DynColors::Rgb(255, 0, 0))
                .to_string()
        ));

        assert!(churn_info
            .value()
            .contains(&"file_2.txt 30".color(DynColors::Rgb(255, 0, 0)).to_string()));
    }

    #[test]
    fn test_truncate_file_path() {
        assert_eq!(shorten_file_path("path/to/file.txt", 3), "path/to/file.txt");
        assert_eq!(shorten_file_path("another/file.txt", 2), "another/file.txt");
        assert_eq!(shorten_file_path("file.txt", 1), "file.txt");
        assert_eq!(
            shorten_file_path("path/to/file.txt", 2),
            "\u{2026}/to/file.txt"
        );
        assert_eq!(
            shorten_file_path("another/file.txt", 1),
            "\u{2026}/file.txt"
        );
        assert_eq!(shorten_file_path("file.txt", 0), "file.txt");
    }
}
