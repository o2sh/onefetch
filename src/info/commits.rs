use serde::Serialize;

use crate::{
    cli::NumberSeparator,
    info::{format_number, utils::git::CommitMetrics, utils::info_field::InfoField},
};

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CommitsInfo {
    pub number_of_commits: usize,
    is_shallow: bool,
    #[serde(skip_serializing)]
    number_separator: NumberSeparator,
}

impl CommitsInfo {
    pub fn new(commit_metrics: &CommitMetrics, number_separator: NumberSeparator) -> Self {
        Self {
            number_of_commits: commit_metrics.total_number_of_commits,
            is_shallow: commit_metrics.is_shallow,
            number_separator,
        }
    }
}

#[typetag::serialize]
impl InfoField for CommitsInfo {
    fn value(&self) -> String {
        format!(
            "{}{}",
            format_number(&self.number_of_commits, self.number_separator),
            self.is_shallow.then_some(" (shallow)").unwrap_or_default()
        )
    }

    fn title(&self) -> String {
        "Commits".into()
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_display_commits_info() {
        let commits_info = CommitsInfo {
            number_of_commits: 3,
            is_shallow: false,
            number_separator: NumberSeparator::Plain,
        };

        assert_eq!(commits_info.value(), "3".to_string());
    }

    #[test]
    fn test_display_commits_info_shallow() {
        let commits_info = CommitsInfo {
            number_of_commits: 2,
            is_shallow: true,
            number_separator: NumberSeparator::Plain,
        };

        assert_eq!(commits_info.value(), "2 (shallow)".to_string());
    }
}
