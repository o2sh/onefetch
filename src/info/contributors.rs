use super::git::metrics::GitMetrics;
use crate::{
    cli::NumberSeparator,
    info::{format_number, utils::info_field::InfoField},
};
use serde::Serialize;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ContributorsInfo {
    pub total_number_of_authors: usize,
    #[serde(skip_serializing)]
    pub number_of_authors_to_display: usize,
    #[serde(skip_serializing)]
    number_separator: NumberSeparator,
}

impl ContributorsInfo {
    pub fn new(
        git_metrics: &GitMetrics,
        number_of_authors_to_display: usize,
        number_separator: NumberSeparator,
    ) -> Self {
        Self {
            total_number_of_authors: git_metrics.total_number_of_authors,
            number_of_authors_to_display,
            number_separator,
        }
    }
}

#[typetag::serialize]
impl InfoField for ContributorsInfo {
    fn value(&self) -> String {
        if self.total_number_of_authors > self.number_of_authors_to_display {
            format_number(&self.total_number_of_authors, self.number_separator)
        } else {
            "".to_string()
        }
    }

    fn title(&self) -> String {
        "Contributors".into()
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use gix::actor::Time;

    #[test]
    fn test_display_contributors_info() {
        let timestamp = Time::now_utc();
        let git_metrics = GitMetrics {
            authors_to_display: vec![],
            file_churns_to_display: vec![],
            total_number_of_authors: 12,
            total_number_of_commits: 2,
            churn_pool_size: 0,
            time_of_most_recent_commit: timestamp,
            time_of_first_commit: timestamp,
        };

        let contributors_info = ContributorsInfo::new(&git_metrics, 2, NumberSeparator::Plain);
        assert_eq!(contributors_info.value(), "12".to_string());
        assert_eq!(contributors_info.title(), "Contributors".to_string());
    }

    #[test]
    fn test_display_contributors_less_than_authors_to_display() {
        let contributors_info = ContributorsInfo {
            total_number_of_authors: 1,
            number_of_authors_to_display: 3,
            number_separator: NumberSeparator::Plain,
        };

        assert!(contributors_info.value().is_empty());
    }
}
