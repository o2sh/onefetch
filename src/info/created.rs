use super::{git::metrics::GitMetrics, utils::format_time};
use crate::info::utils::info_field::InfoField;
use serde::Serialize;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CreatedInfo {
    pub creation_date: String,
}

impl CreatedInfo {
    pub fn new(iso_time: bool, git_metrics: &GitMetrics) -> Self {
        let creation_date = get_creation_date(git_metrics, iso_time);
        Self { creation_date }
    }
}

fn get_creation_date(git_metrics: &GitMetrics, iso_time: bool) -> String {
    format_time(git_metrics.time_of_first_commit, iso_time)
}

#[typetag::serialize]
impl InfoField for CreatedInfo {
    fn value(&self) -> String {
        self.creation_date.to_string()
    }

    fn title(&self) -> String {
        "Created".into()
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_display_created_info() {
        let created_info = CreatedInfo {
            creation_date: "2 years ago".to_string(),
        };

        assert_eq!(created_info.value(), "2 years ago".to_string());
    }
}
