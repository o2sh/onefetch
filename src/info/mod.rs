use self::authors::AuthorsInfo;
use self::churn::ChurnInfo;
use self::commits::CommitsInfo;
use self::contributors::ContributorsInfo;
use self::created::CreatedInfo;
use self::dependencies::DependenciesInfo;
use self::description::DescriptionInfo;
use self::git::metrics::GitMetrics;
use self::git::traverse_commit_graph;
use self::head::HeadInfo;
use self::langs::language::Language;
use self::langs::language::LanguagesInfo;
use self::last_change::LastChangeInfo;
use self::license::LicenseInfo;
use self::loc::LocInfo;
use self::pending::PendingInfo;
use self::project::ProjectInfo;
use self::size::SizeInfo;
use self::title::Title;
use self::url::get_repo_url;
use self::url::UrlInfo;
use self::utils::info_field::{InfoField, InfoType};
use self::version::VersionInfo;
use crate::cli::{is_truecolor_terminal, CliOptions, NumberSeparator, When};
use crate::ui::get_ascii_colors;
use crate::ui::text_colors::TextColors;
use anyhow::{Context, Result};
use gix::sec::trust::Mapping;
use gix::Repository;
use onefetch_manifest::Manifest;
use owo_colors::{DynColors, OwoColorize};
use serde::Serialize;
use std::path::Path;

mod authors;
mod churn;
mod commits;
mod contributors;
mod created;
mod dependencies;
mod description;
mod git;
mod head;
pub mod langs;
mod last_change;
mod license;
mod loc;
mod pending;
mod project;
mod size;
mod title;
mod url;
pub mod utils;
mod version;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Info {
    title: Option<Title>,
    info_fields: Vec<Box<dyn InfoField>>,
    #[serde(skip_serializing)]
    text_colors: TextColors,
    #[serde(skip_serializing)]
    no_color_palette: bool,
    #[serde(skip_serializing)]
    no_bold: bool,
    #[serde(skip_serializing)]
    pub dominant_language: Language,
    #[serde(skip_serializing)]
    pub ascii_colors: Vec<DynColors>,
}

struct InfoBuilder {
    title: Option<Title>,
    info_fields: Vec<Box<dyn InfoField>>,
    disabled_fields: Vec<InfoType>,
    no_title: bool,
}

impl std::fmt::Display for Info {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        //Title
        if let Some(title) = &self.title {
            write!(f, "{title}")?;
        }

        //Info lines
        for info_field in self.info_fields.iter() {
            info_field.write_styled(f, self.no_bold, &self.text_colors)?;
        }

        //Palette
        if !self.no_color_palette {
            writeln!(
                f,
                "\n{0}{1}{2}{3}{4}{5}{6}{7}",
                "   ".on_black(),
                "   ".on_red(),
                "   ".on_green(),
                "   ".on_yellow(),
                "   ".on_blue(),
                "   ".on_magenta(),
                "   ".on_cyan(),
                "   ".on_white()
            )?;
        }

        Ok(())
    }
}

pub fn build_info(cli_options: &CliOptions) -> Result<Info> {
    let mut repo = gix::ThreadSafeRepository::discover_opts(
        &cli_options.input,
        gix::discover::upwards::Options {
            dot_git_only: true,
            ..Default::default()
        },
        Mapping::default(),
    )?
    .to_thread_local();
    // Having an object cache is important for getting much better traversal and diff performance.
    repo.object_cache_size_if_unset(4 * 1024 * 1024);
    let repo_path = get_work_dir(&repo)?;

    let loc_by_language_sorted_handle = std::thread::spawn({
        let globs_to_exclude = cli_options.info.exclude.clone();
        let language_types = cli_options.info.r#type.clone();
        let include_hidden = cli_options.info.include_hidden;
        let workdir = repo_path.clone();
        move || {
            langs::get_loc_by_language_sorted(
                &workdir,
                &globs_to_exclude,
                &language_types,
                include_hidden,
            )
        }
    });

    let loc_by_language = loc_by_language_sorted_handle
        .join()
        .ok()
        .context("BUG: panic in language statistics thread")??;
    let manifest = get_manifest(&repo_path)?;
    let repo_url = get_repo_url(
        &repo,
        cli_options.info.hide_token,
        cli_options.info.http_url,
    );

    let git_metrics = traverse_commit_graph(
        &repo,
        cli_options.info.no_bots.clone(),
        cli_options.info.churn_pool_size,
        cli_options.info.no_merges,
    )?;
    let true_color = match cli_options.ascii.true_color {
        When::Always => true,
        When::Never => false,
        When::Auto => is_truecolor_terminal(),
    };
    let dominant_language = langs::get_main_language(&loc_by_language);
    let ascii_colors = get_ascii_colors(
        cli_options.ascii.ascii_language.as_ref(),
        &dominant_language,
        &cli_options.ascii.ascii_colors,
        true_color,
    );

    let text_colors = TextColors::new(&cli_options.text_formatting.text_colors, ascii_colors[0]);
    let no_bold = cli_options.text_formatting.no_bold;
    let number_separator = cli_options.text_formatting.number_separator;
    let iso_time = cli_options.text_formatting.iso_time;
    let number_of_languages_to_display = cli_options.info.number_of_languages;
    let number_of_authors_to_display = cli_options.info.number_of_authors;
    let number_of_file_churns_to_display = cli_options.info.number_of_file_churns;
    let globs_to_exclude = &cli_options.info.exclude;
    let show_email = cli_options.info.email;

    Ok(InfoBuilder::new(cli_options)
        .title(&repo, no_bold, &text_colors)
        .project(&repo, &repo_url, manifest.as_ref(), number_separator)?
        .description(manifest.as_ref())
        .head(&repo)?
        .pending(&repo)?
        .version(&repo, manifest.as_ref())?
        .created(&git_metrics, iso_time)
        .languages(
            &loc_by_language,
            true_color,
            number_of_languages_to_display,
            &text_colors,
            cli_options,
        )
        .dependencies(manifest.as_ref(), number_separator)
        .authors(
            &git_metrics,
            number_of_authors_to_display,
            show_email,
            number_separator,
        )
        .last_change(&git_metrics, iso_time)
        .contributors(&git_metrics, number_of_authors_to_display, number_separator)
        .url(&repo_url)
        .commits(&git_metrics, repo.is_shallow(), number_separator)
        .churn(
            &git_metrics,
            number_of_file_churns_to_display,
            globs_to_exclude,
            number_separator,
        )?
        .loc(&loc_by_language, number_separator)
        .size(&repo, number_separator)
        .license(&repo_path, manifest.as_ref())?
        .build(cli_options, text_colors, dominant_language, ascii_colors))
}

impl InfoBuilder {
    fn new(cli_options: &CliOptions) -> Self {
        Self {
            title: None,
            info_fields: Vec::new(),
            disabled_fields: cli_options.info.disabled_fields.clone(),
            no_title: cli_options.info.no_title,
        }
    }

    fn title(mut self, repo: &Repository, no_bold: bool, text_colors: &TextColors) -> Self {
        if !self.no_title {
            let title = Title::new(
                repo,
                text_colors.title,
                text_colors.tilde,
                text_colors.underline,
                !no_bold,
            );
            self.title = Some(title);
        }
        self
    }

    fn description(mut self, manifest: Option<&Manifest>) -> Self {
        if !self.disabled_fields.contains(&InfoType::Description) {
            let description = DescriptionInfo::new(manifest);
            self.info_fields.push(Box::new(description));
        }
        self
    }

    fn pending(mut self, repo: &Repository) -> Result<Self> {
        if !self.disabled_fields.contains(&InfoType::Pending) {
            let pending = PendingInfo::new(repo)?;
            self.info_fields.push(Box::new(pending));
        }
        Ok(self)
    }

    fn url(mut self, repo_url: &str) -> Self {
        if !self.disabled_fields.contains(&InfoType::URL) {
            let repo_url = UrlInfo::new(repo_url);
            self.info_fields.push(Box::new(repo_url));
        }
        self
    }

    fn project(
        mut self,
        repo: &Repository,
        repo_url: &str,
        manifest: Option<&Manifest>,
        number_separator: NumberSeparator,
    ) -> Result<Self> {
        if !self.disabled_fields.contains(&InfoType::Project) {
            let project = ProjectInfo::new(repo, repo_url, manifest, number_separator)?;
            self.info_fields.push(Box::new(project));
        }
        Ok(self)
    }

    fn head(mut self, repo: &Repository) -> Result<Self> {
        if !self.disabled_fields.contains(&InfoType::Head) {
            let head = HeadInfo::new(repo)?;
            self.info_fields.push(Box::new(head));
        }
        Ok(self)
    }

    fn version(mut self, repo: &Repository, manifest: Option<&Manifest>) -> Result<Self> {
        if !self.disabled_fields.contains(&InfoType::Version) {
            let version = VersionInfo::new(repo, manifest)?;
            self.info_fields.push(Box::new(version));
        }
        Ok(self)
    }

    fn size(mut self, repo: &Repository, number_separator: NumberSeparator) -> Self {
        if !self.disabled_fields.contains(&InfoType::Size) {
            let size = SizeInfo::new(repo, number_separator);
            self.info_fields.push(Box::new(size));
        }
        self
    }

    fn license(mut self, repo_path: &Path, manifest: Option<&Manifest>) -> Result<Self> {
        if !self.disabled_fields.contains(&InfoType::License) {
            let license = LicenseInfo::new(repo_path, manifest)?;
            self.info_fields.push(Box::new(license));
        }
        Ok(self)
    }

    fn created(mut self, git_metrics: &GitMetrics, iso_time: bool) -> Self {
        if !self.disabled_fields.contains(&InfoType::Created) {
            let created = CreatedInfo::new(iso_time, git_metrics);
            self.info_fields.push(Box::new(created));
        }
        self
    }

    fn languages(
        mut self,
        loc_by_language: &[(Language, usize)],
        true_color: bool,
        number_of_languages: usize,
        text_colors: &TextColors,
        cli_options: &CliOptions,
    ) -> Self {
        if !self.disabled_fields.contains(&InfoType::Languages) {
            let languages = LanguagesInfo::new(
                loc_by_language,
                true_color,
                number_of_languages,
                text_colors.info,
                cli_options.visuals.nerd_fonts,
            );
            self.info_fields.push(Box::new(languages));
        }
        self
    }

    fn dependencies(
        mut self,
        manifest: Option<&Manifest>,
        number_separator: NumberSeparator,
    ) -> Self {
        if !self.disabled_fields.contains(&InfoType::Dependencies) {
            let dependencies = DependenciesInfo::new(manifest, number_separator);
            self.info_fields.push(Box::new(dependencies));
        }
        self
    }

    fn authors(
        mut self,
        git_metrics: &GitMetrics,
        number_of_authors_to_display: usize,
        show_email: bool,
        number_separator: NumberSeparator,
    ) -> Self {
        if !self.disabled_fields.contains(&InfoType::Authors) {
            let authors = AuthorsInfo::new(
                &git_metrics.number_of_commits_by_signature,
                git_metrics.total_number_of_commits,
                number_of_authors_to_display,
                show_email,
                number_separator,
            );
            self.info_fields.push(Box::new(authors));
        }
        self
    }

    fn last_change(mut self, git_metrics: &GitMetrics, iso_time: bool) -> Self {
        if !self.disabled_fields.contains(&InfoType::LastChange) {
            let last_change = LastChangeInfo::new(iso_time, git_metrics);
            self.info_fields.push(Box::new(last_change));
        }
        self
    }

    fn contributors(
        mut self,
        git_metrics: &GitMetrics,
        number_of_authors_to_display: usize,
        number_separator: NumberSeparator,
    ) -> Self {
        if !self.disabled_fields.contains(&InfoType::Contributors) {
            let contributors = ContributorsInfo::new(
                git_metrics.total_number_of_authors,
                number_of_authors_to_display,
                number_separator,
            );
            self.info_fields.push(Box::new(contributors));
        }
        self
    }

    fn commits(
        mut self,
        git_metrics: &GitMetrics,
        is_shallow: bool,
        number_separator: NumberSeparator,
    ) -> Self {
        if !self.disabled_fields.contains(&InfoType::Commits) {
            let commits = CommitsInfo::new(git_metrics, is_shallow, number_separator);
            self.info_fields.push(Box::new(commits));
        }
        self
    }

    fn churn(
        mut self,
        git_metrics: &GitMetrics,
        number_of_file_churns_to_display: usize,
        globs_to_exclude: &[String],
        number_separator: NumberSeparator,
    ) -> Result<Self> {
        if !self.disabled_fields.contains(&InfoType::Churn) {
            let churn = ChurnInfo::new(
                &git_metrics.number_of_commits_by_file_path,
                git_metrics.churn_pool_size,
                number_of_file_churns_to_display,
                globs_to_exclude,
                number_separator,
            )?;
            self.info_fields.push(Box::new(churn));
        }
        Ok(self)
    }

    fn loc(
        mut self,
        loc_by_language: &[(Language, usize)],
        number_separator: NumberSeparator,
    ) -> Self {
        if !self.disabled_fields.contains(&InfoType::LinesOfCode) {
            let lines_of_code = LocInfo::new(loc_by_language, number_separator);
            self.info_fields.push(Box::new(lines_of_code));
        }
        self
    }

    fn build(
        self,
        cli_options: &CliOptions,
        text_colors: TextColors,
        dominant_language: Language,
        ascii_colors: Vec<DynColors>,
    ) -> Info {
        Info {
            title: self.title,
            info_fields: self.info_fields,
            text_colors,
            dominant_language,
            ascii_colors,
            no_color_palette: cli_options.visuals.no_color_palette,
            no_bold: cli_options.text_formatting.no_bold,
        }
    }
}

fn get_manifest(repo_path: &Path) -> Result<Option<Manifest>> {
    let manifests = onefetch_manifest::get_manifests(repo_path)?;

    if manifests.is_empty() {
        Ok(None)
    } else {
        Ok(manifests.first().cloned())
    }
}

pub fn get_work_dir(repo: &gix::Repository) -> Result<std::path::PathBuf> {
    Ok(repo
        .work_dir()
        .context("please run onefetch inside of a non-bare git repository")?
        .to_owned())
}
