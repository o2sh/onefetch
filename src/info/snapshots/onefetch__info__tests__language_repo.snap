---
source: src/info/mod.rs
expression: info
---
{
  "gitUsername": "onefetch-committer-name",
  "gitVersion": "git version",
  "project": {
    "repo_name": "repo",
    "number_of_branches": 0,
    "number_of_tags": 1
  },
  "head": {
    "short_commit_id": "short commit",
    "refs": [
      "main"
    ]
  },
  "version": "tag1",
  "created": "22 years ago",
  "languages": [
    {
      "language": "Verilog",
      "percentage": 100.0
    }
  ],
  "dependencies": "1 (cargo)",
  "authors": [
    {
      "name": "Author Four",
      "email": "author4@example.org",
      "nbr_of_commits": 1,
      "contribution": 25
    },
    {
      "name": "Author One",
      "email": "author1@example.org",
      "nbr_of_commits": 1,
      "contribution": 25
    },
    {
      "name": "Author Three",
      "email": "author3@example.org",
      "nbr_of_commits": 1,
      "contribution": 25
    }
  ],
  "lastChange": "22 years ago",
  "contributors": 4,
  "repoUrl": "https://github.com/user/repo.git",
  "numberOfCommits": "4",
  "linesOfCode": 4,
  "repoSize": "22 B",
  "license": "MIT"
}
